src/main/scala/rise/core/Expr.scala

```diff
@@ -1,46 +1,82 @@
 package rise.core
 
+import semantics._
 import rise.core.types._
 import rise.core.DSL.Type.TypeEqual
 import rise.core.ShowRise._
 
+object alphaEquiv {
+  type TYPEEQ = Type => Type => Boolean
+
+  val equiv : TYPEEQ => Expr => Expr => Boolean = typeEq => a => b => {
+    typeEq(a.t)(b.t) && ((a, b) match {
+      case (Identifier(na), Identifier(nb)) => na == nb
+      case (Literal(da), Literal(db)) => da == db
+      case (a : Primitive, b : Primitive) => a.name == b.name
+      // Application compares structurally
+      case (App(fa, ea), App(fb, eb)) => equiv(typeEq)(fa)(fb) && equiv(typeEq)(ea)(eb)
+      case (DepApp(fa, xa), DepApp(fb, xb)) => equiv(typeEq)(fa)(fb) && xa == xb
+      // Abstraction compares after substitution
+      case (Lambda(xa, ta), other@Lambda(xb, _)) =>
+        typeEq(xa.t)(xb.t) && equiv(typeEq)(ta)(typedLifting.liftFunExpr(other).value(xa))
+      case (DepLambda(xa, ea), other@DepLambda(xb, _)) => (xa, xb) match {
+        case (n: NatIdentifier, _: NatIdentifier) =>
+          equiv(typeEq)(ea)(typedLifting.liftDepFunExpr[NatKind](other).value(n))
+        case (dt: DataTypeIdentifier, _: DataTypeIdentifier) =>
+          equiv(typeEq)(ea)(typedLifting.liftDepFunExpr[DataKind](other).value(dt))
+        case (addr: AddressSpaceIdentifier, _: AddressSpaceIdentifier) =>
+          equiv(typeEq)(ea)(typedLifting.liftDepFunExpr[AddressSpaceKind](other).value(addr))
+        case (n2n: NatToNatIdentifier, _: NatToNatIdentifier) =>
+          equiv(typeEq)(ea)(typedLifting.liftDepFunExpr[NatToNatKind](other).value(n2n))
+        case (n2d: NatToDataIdentifier, _: NatToDataIdentifier) =>
+          equiv(typeEq)(ea)(typedLifting.liftDepFunExpr[NatToDataKind](other).value(n2d))
+        case _ => false
+      }
+      case _ => false
+    })
+  }
+
+  val hash : Expr => Int = {
+    case _: Identifier => 17
+    case Lambda(_, e) => 3 * e.hashCode() + 1
+    case App(f, e) => 5 * f.hashCode() + -7 * e.hashCode() + 2
+    case DepLambda(_, e) => 4 * e.hashCode() + 3
+    case DepApp(f, _) => 6 * f.hashCode() + 4
+    case l@Literal(_: ScalarData | _: VectorData) => l.d.hashCode()
+    case Literal(_: NatData) => 91
+    case Literal(_: IndexData) => 93
+    case Literal(_: ArrayData) => 95
+    case Literal(_: PairData) => 97
+    case p: Primitive => p.getClass.hashCode()
```

Comment:
`p.name.hashCode()` might allow to differentiate between e.g. `mapLocal(1)` and `mapLocal(2)`?

Reply:
Didn't know about those, thanks!